File: PHP/Car/Point.php
<?php
    require_once 'Exceptions/InvalidParam.php';
    require_once 'Exceptions/InvalidAttribute.php';
    
    class Point {
        private $x, $y; 
        
        protected function validate($value) {
            if ( is_numeric($value) ) {
                return $value;
            }
            throw new InvalidParam("Invalid param '$value'.");
        }
        
        public function __construct($x = 0.0, $y = 0.0) {
            $this->x = $this->validate($x);
            $this->y = $this->validate($y);
        }
        
        public function __get($name) {
            if ( property_exists($this, $name) ) {
                return $this->$name;    
            }
            throw new InvalidAttribute("Attribute error: attribute '$name' not found.");
        }
        
        public function distance(Point $other) {
            return hypot($other->x-$this->x, $other->y-$this->y);
        }
        
        public function compare(Point $other) {
            if ( $this == $other ) {
                echo $this . " is equeal to " . $other . PHP_EOL;
            } else {
                echo $this . " is not equal to " . $other . PHP_EOL;
            }
        }
        
        public function __toString() {
            return "($this->x, $this->y)";
        }
    }
?>

File: W5/Car/Point.py
import math


class Point:
    def __init__(self, x: float = 0.0, y: float = 0.0) -> None:
        self._x = float(x)
        self._y = float(y)

    def __str__(self) -> str:
        return f'({self.x}, {self.y})'

    def __repr__(self) -> str:
        return f'({self.x}, {self.y})'

    def __eq__(self, other: 'Point') -> bool:
        return self.x == other.x and self.y == other.y

    def __ne__(self, other: 'Point') -> bool:
        return not self == other

    @property
    def x(self) -> float:
        return self._x

    @property
    def y(self) -> float:
        return self._y

    def distance(self, other: 'Point') -> float:
        return math.hypot(other.x - self.x, other.y - self.y)
